server/modules/protocol/MySQL/MySQLBackend/mysql_backend.c

```diff
@@ -895,6 +895,7 @@ gw_read_and_write(DCB *dcb)
         if (session_ok_to_route(dcb))
         {
             gwbuf_set_type(stmt, GWBUF_TYPE_MYSQL);
+            stmt->server = dcb->server;
```

Comment:
better place to assign this pointer would be in `dcb_read` where the buffer is allocated. The DCBs have a SERVER pointer which could always be assigned to the buffer since the client DCB's SERVER pointer is NULL.

Look at line 870 in `server/core/dcb.c`, I think this would be a good place to do it.